app/services/catalog/folio_writer.rb

```diff
@@ -0,0 +1,78 @@
+# frozen_string_literal: true
+
+module Catalog
+  # Updates the FOLIO MARC record's 856 fields
+  class FolioWriter
+    def self.save(cocina_object:, marc_856_data:)
+      new(cocina_object:, marc_856_data:).save
+    end
+
+    def initialize(cocina_object:, marc_856_data:)
+      @cocina_object = cocina_object
+      @marc_856_data = marc_856_data
+    end
+
+    # Determines which Folio records to update and handles adding / deleting 856 fields
+    # @raise [FolioClient::Error] FolioClient base error if problem interacting with Folio
+    def save
+      return if catalog_record_ids.empty? && previous_catalog_record_ids.empty?
+
+      # remove 856 for previous catkeys
+      previous_catalog_record_ids.each { |previous_id| delete_previous_ids(catalog_record_id: previous_id) }
+
+      # replace 856 for current catkeys
+      catalog_record_ids.each { |catalog_record_id| update_current_ids(catalog_record_id:) }
+    end
+
+    private
+
+    attr_reader :marc_856_data, :cocina_object
+
+    def delete_previous_ids(catalog_record_id:)
+      FolioClient.edit_marc_json(hrid: catalog_record_id) do |marc_json|
+        marc_json['fields'].reject! { |field| (field['tag'] == '856') && (field['content'].include? purl_subfield) }
+      end
+    end
+
+    def update_current_ids(catalog_record_id:)
+      FolioClient.edit_marc_json(hrid: catalog_record_id) do |marc_json|
+        marc_json['fields'].reject! { |field| (field['tag'] == '856') && (field['content'].include? purl_subfield) }
+        marc_json['fields'] << marc_856_field if released_to_searchworks?(cocina_object)
+      end
+    end
+
+    def purl_subfield
+      "$u #{cocina_object.description.purl}"
+    end
+
+    def catalog_record_ids
+      @catalog_record_ids ||= fetch_catalog_record_ids(current: true)
+    end
+
+    def previous_catalog_record_ids
+      @previous_catalog_record_ids ||= fetch_catalog_record_ids(current: false)
+    end
+
+    # List of current or previous catalog record ids for the cocina object (depending on parameter passed)
+    # @param current [boolean] if you want the current or previous catalog record ids
+    # @return [Array] previous or current catalog_record_ids for the object in an array, empty array if none exist
+    def fetch_catalog_record_ids(current:)
+      ckey_type = current ? 'folio' : 'previous folio'
+      @cocina_object.identification.catalogLinks.select { |link| link.catalog == ckey_type }.map(&:catalogRecordId)
```

Comment:
What would you think about using `filter_map` in lieu of `select.map` to get down to a single iteration?

```suggestion
      ckey_type = current ? 'folio' : 'previous folio'
      @cocina_object.identification.catalogLinks.filter_map { |link| link.catalogRecordId if link.catalog == ckey_type }
```

Reply:
Yes, that's more concise and makes sense. Have updated that.